hydroflow_plus/src/persist_pushdown.rs

```diff
@@ -0,0 +1,71 @@
+use crate::ir::*;
+
+fn persist_pullup_node(node: HfPlusNode, seen_tees: &mut SeenTees) -> HfPlusNode {
+    match node.transform_children(persist_pullup_node, seen_tees) {
+        HfPlusNode::Map {
+            f,
+            input: box HfPlusNode::Persist(behind_persist),
+        } => HfPlusNode::Persist(Box::new(HfPlusNode::Map {
+            f,
+            input: behind_persist,
+        })),
+        o => o,
+    }
+}
+
+pub fn persist_pullup(ir: Vec<HfPlusLeaf>) -> Vec<HfPlusLeaf> {
+    let mut seen_tees = Default::default();
+    ir.into_iter()
+        .map(|l| l.transform_children(persist_pullup_node, &mut seen_tees))
+        .collect()
+}
```

Comment:
would it make sense to combine these two types of things into one trait?

Reply:
I think I want to wait a bit till we write some optimizations, cause I'm not super sure what patterns might arise. Like it may make sense to have the trait include traversal logic too.